src/lib/components/Swap/SwapButton.tsx

```diff
@@ -61,72 +59,73 @@ export default function SwapButton({ disabled }: SwapButtonProps) {
-    useIsPendingApproval,
-    approvalCurrencyAmount
-  )
+
+  const { notApproved, pendingApproval, loadingSignature, supportsPermit, signatureData, handleApproveOrPermit } =
+    useApproveAndPermit(optimizedTrade, allowedSlippage, useIsPendingApproval, approvalCurrencyAmount)
+
   const approvalHash = usePendingApproval(
     inputCurrency?.isToken ? inputCurrency : undefined,
     useSwapRouterAddress(optimizedTrade)
   )
 
   const addTransaction = useAddTransaction()
-  const addApprovalTransaction = useCallback(() => {
-    getApproval().then((transaction) => {
-      if (transaction) {
-        addTransaction({ type: TransactionType.APPROVAL, ...transaction })
-      }
-    })
-  }, [addTransaction, getApproval])
+  const deadline = useTransactionDeadline()
 
   const actionProps = useMemo((): Partial<ActionButtonProps> | undefined => {
-    if (!disabled && chainId) {
-      if (approval === ApprovalState.NOT_APPROVED) {
-        const currency = inputCurrency || approvalCurrencyAmount?.currency
-        invariant(currency)
-        return {
-          action: {
-            message: <Trans>Approve {currency.symbol} first</Trans>,
-            onClick: addApprovalTransaction,
-            children: <Trans>Approve</Trans>,
-          },
-        }
-      } else if (approval === ApprovalState.PENDING) {
-        return {
-          disabled: true,
-          action: {
-            message: (
-              <EtherscanLink type={ExplorerDataType.TRANSACTION} data={approvalHash}>
-                <Trans>Approval pending</Trans>
-              </EtherscanLink>
-            ),
-            icon: Spinner,
-            onClick: addApprovalTransaction,
-            children: <Trans>Approve</Trans>,
-          },
-        }
-      } else if (inputCurrencyAmount && inputCurrencyBalance && !inputCurrencyBalance.lessThan(inputCurrencyAmount)) {
-        return {}
+    if (
+      disabled ||
+      !chainId ||
+      loadingSignature ||
+      (inputCurrencyAmount && inputCurrencyBalance && inputCurrencyBalance.lessThan(inputCurrencyAmount))
+    ) {
+      return { disabled: true }
+    }
+    if (notApproved) {
+      const currency = inputCurrency || approvalCurrencyAmount?.currency
+      invariant(currency)
+      return {
+        action: {
+          message: supportsPermit ? (
+            <Trans>Allow {currency.symbol} first</Trans>
+          ) : (
+            <Trans>Approve {currency.symbol} first</Trans>
+          ),
+          onClick: handleApproveOrPermit,
+          children: supportsPermit ? <Trans>Allow</Trans> : <Trans>Approve</Trans>,
+        },
+      }
+    }
+    if (pendingApproval) {
+      return {
+        disabled: true,
+        action: {
+          message: (
+            <EtherscanLink type={ExplorerDataType.TRANSACTION} data={approvalHash}>
+              <Trans>Approval pending</Trans>
+            </EtherscanLink>
+          ),
+          icon: Spinner,
+          onClick: handleApproveOrPermit,
```

Comment:
omit this property. if typings do not allow it, make it a nop. (Jordan is updating typings to allow it.)